<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<HTML>
 <HEAD>
   <TITLE> [c++-pthreads] What are the real issues?
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:c%2B%2B-pthreads%40codesourcery.com?Subject=Re%3A%20%5Bc%2B%2B-pthreads%5D%20What%20are%20the%20real%20issues%3F&In-Reply-To=%3Cxyp7k04vja2.fsf%40miranda.boston.redhat.com%3E">
   <META NAME="robots" CONTENT="index,nofollow">
   <style type="text/css">
       pre {
           white-space: pre-wrap;       /* css-2.1, curent FF, Opera, Safari */
           }
   </style>
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="000106.html">
   <LINK REL="Next"  HREF="000137.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>[c++-pthreads] What are the real issues?</H1>
    <B>Jason Merrill</B> 
    <A HREF="mailto:c%2B%2B-pthreads%40codesourcery.com?Subject=Re%3A%20%5Bc%2B%2B-pthreads%5D%20What%20are%20the%20real%20issues%3F&In-Reply-To=%3Cxyp7k04vja2.fsf%40miranda.boston.redhat.com%3E"
       TITLE="[c++-pthreads] What are the real issues?">jason at redhat.com
       </A><BR>
    <I>Tue Jan  6 21:50:29 UTC 2004</I>
    <P><UL>
        <LI>Previous message: <A HREF="000106.html">[c++-pthreads] What are the real issues?
</A></li>
        <LI>Next message: <A HREF="000137.html">[c++-pthreads] What are the real issues?
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#107">[ date ]</a>
              <a href="thread.html#107">[ thread ]</a>
              <a href="subject.html#107">[ subject ]</a>
              <a href="author.html#107">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>On Tue, 6 Jan 2004 13:29:22 -0800, Matt Austern &lt;<A HREF="http://sourcerytools.com/cgi-bin/mailman/listinfo/c++-pthreads">austern at apple.com</A>&gt; wrote:

&gt;<i> On Jan 6, 2004, at 1:12 PM, Nathan Myers wrote:
</I>&gt;<i>
</I>&gt;&gt;&gt;<i> ... I think the really fundamental issue is whether a thread
</I>&gt;&gt;&gt;<i> should be allowed to receive a cancellation request, start to do some
</I>&gt;&gt;&gt;<i> work as a result of the request, and then decide that it doesn't want
</I>&gt;&gt;&gt;<i> to be cancelled.
</I>&gt;&gt;<i>
</I>&gt;&gt;<i> Nobody has proposed any such thing.
</I>&gt;<i>
</I>&gt;<i> Really?  I think that you and I both did.
</I>&gt;<i>   (1) My proposal (which may or may not be a good idea, but which
</I>&gt;<i>         I think should be considered) is that cancellation could be
</I>&gt;<i>         represented by an ordinary exception instead of by a special
</I>&gt;<i>         kind of exception with forced unwinding semantics.  This would
</I>&gt;<i>         allow a thread to catch the exception, set a flag, swallow the
</I>&gt;<i>         exception, and continue executing.  that is, it would allow a
</I>&gt;<i>         thread to receive and act on a cancellation request but for it not
</I>&gt;<i>         to be canceled.
</I>
My proposal is that it acts as a normal exception, but that

 * If a cancellation exception is destroyed, the cancellation request
   is re-entered, and acted on again at the next cancellation point.

So the cancellation request never goes away until the thread does.

&gt;<i>   (2) Your proposal is that the POSIX functions that are currently
</I>&gt;<i>         described as cancellation points shouldn't raise any kind of
</I>&gt;<i>         exception and shouldn't cause the thread to stop executing,
</I>&gt;<i>         but should instead return an error code.  Again, this would allow
</I>&gt;<i>         a thread to note the error return, check errno to see if it's
</I>&gt;<i>         ETHREADCANCELED, set a flag as a result of having received
</I>&gt;<i>         the cancelation request, and continue executing.  Again, your
</I>&gt;<i>         proposal would allow a thread to receive and act on a cancellation
</I>&gt;<i>         request without being canceled.
</I>
You are assuming that this would only happen for the first call to a
cancellation point, which doesn't seem to be what Nathan had in mind.
Here's his proposal again:

  C library and system calls never throw.
  In the event of cancellation, cancellation point functions report 
    failure via their normal means (e.g. return -1, or NULL), and set 
    errno if appropriate.  Blocked calls fail immediately.
  C++ library functions that are normally allowed to throw may throw
    a cancellation exception.  (E.g. std::filebuf::overflow().)
  A thread, once cancelled, stays cancelled, regardless of any C++ 
    catch blocks entered or left.
  C++ catch blocks work normally; a cancellation exception may be
    rethrown, or swallowed like any other. 
  In a catch block handling a cancellation exception, functions
    identified as cancellation points work normally.
  When a cancellation exception is swallowed by a catch block, 
    subsequently-called cancellation point functions report failure, 
    as before, until another cancellation exception is thrown.
  Eventually the inability of the thread to achieve anything leads it
    to top-level code equipped to recognize the cancelled condition as 
    such, and to clean up and die in a controlled way.

Jason

</PRE>

<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="000106.html">[c++-pthreads] What are the real issues?
</A></li>
	<LI>Next message: <A HREF="000137.html">[c++-pthreads] What are the real issues?
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#107">[ date ]</a>
              <a href="thread.html#107">[ thread ]</a>
              <a href="subject.html#107">[ subject ]</a>
              <a href="author.html#107">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="http://sourcerytools.com/cgi-bin/mailman/listinfo/c++-pthreads">More information about the c++-pthreads
mailing list</a><br>
</body></html>
